### PR TITLE
Auto pruners

### PR DESCRIPTION
**Pull request by suiguoxin**
Add algo implementation / examples / test / doc for the following pruning algos:
- NetAdapt
- SimulatedAnnealing
- ADMM
- AutoCompress

### CODE REVIEW

File: docs/en_US/Compressor/Pruner.md [hunk 1]
```diff
@@ -380,5 +385,168 @@ You can view example for more information
...
+```
+
+You can view [example](https://github.com/suiguoxin/nni/tree/auto-pruners/examples/model_compress/auto_pruners_torch.py) for more information.
+
+#### User configuration for AutoCompress Pruner
+
+- **sparsity:** How much percentage of convolutional filters are to be pruned.
+- **op_types:** Currently only Conv2d is supported in TaylorFOWeightFilterPruner.
+- **trainer:** Function used for the first step of ADMM training.
+- **evaluator:** Function to evaluate the masked model.
```
Comment (chicm-ms) [inline docs/en_US/Compressor/Pruner.md:540]: Suggest to describe the requirement, input/output of `trainer` and `evaluator`, and how they are used, better to provide an example.

File: src/sdk/pynni/nni/compression/torch/pruning/admm_pruner.py [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+        wrapper : Module
+            the module to instrument the compression operation
+        Returns
+        -------
+        dict
+            dictionary for storing masks
+        """
+
+        if self._pruning_mode == 'fine_grained':
+            return LevelPruner.calc_mask(self, wrapper, **kwargs)
```
Comment (chicm-ms) [inline src/sdk/pynni/nni/compression/torch/pruning/admm_pruner.py:140]: L1FilterPruner can prune only Conv2d layer, should we check the configuration list and pruning mode together?

File: src/sdk/pynni/nni/compression/torch/pruning/admm_pruner.py [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+        elif self._pruning_mode == 'channel':
+            return L1FilterPruner.calc_mask(self, wrapper, **kwargs)
```
Comment (chicm-ms) [inline src/sdk/pynni/nni/compression/torch/pruning/admm_pruner.py:142]: `return L1FilterPruner.calc_mask(self, wrapper, **kwargs)` 
This is a nice way to integrate one Pruner into another, but it seems only work for those pruners which don't need intialization,  some pruners such as Slim,  Apoz , TalorOF , need initialization before being used.   Do you have requirement to use such pruners  in another Pruner?

Comment (chicm-ms) [reply src/sdk/pynni/nni/compression/torch/pruning/admm_pruner.py:142]: The code directory structure is refactored, the `pruning` directory is planned to place the pure pruning algorithms which do not contain any training schedule. We can create another directory to place the Pruners which contains training schedule such as Anneal, Auto compress etc,   we can discuss it later.

File: src/sdk/pynni/nni/compression/torch/pruning/net_adapt_pruner.py [hunk 4]
```diff
@@ -0,0 +1,410 @@
...
+
+        self._fine_tuner = fine_tuner
+        self._evaluator = evaluator
+        self._optimize_mode = OptimizeMode(optimize_mode)
+
+        # hyper parameters for NetAdapt algorithm
+        self._pruning_step = pruning_step
+
+        # overall pruning rate
+        self._sparsity = config_list[0]['sparsity']
```
Comment (chicm-ms) [inline src/sdk/pynni/nni/compression/torch/pruning/net_adapt_pruner.py:77]: We may need an overall configuration section in our config, especially for those algorithms with training schedule.

File: src/sdk/pynni/nni/compression/torch/pruning/net_adapt_pruner.py [hunk 5]
```diff
@@ -0,0 +1,410 @@
...
+        Returns
+        -------
+        int
+            total weights of all the op considered
+        '''
+        num_weights = 0
+        if count_non_prunable_ops:
+            pruner = LevelPruner(
+                copy.deepcopy(self._model_to_prune), [{'sparsity': 0.1, 'op_types': ['default']}])
+            modules_wrapper = pruner.get_modules_wrapper()
```
Comment (chicm-ms) [inline src/sdk/pynni/nni/compression/torch/pruning/net_adapt_pruner.py:132]: does the deepcopy support all models?

File: src/sdk/pynni/nni/compression/torch/pruning/net_adapt_pruner.py [hunk 6]
```diff
@@ -0,0 +1,410 @@
...
+        if self.modules_to_compress is None:
+            self.modules_to_compress = []
+            # consider only the layers without dependencies
+            model_name = self._model_to_prune.__class__.__name__
+            ops_no_dependency = get_layers_no_dependency(model_name)
+
+            for name, module in self.bound_model.named_modules():
+                if module == self.bound_model:
+                    continue
+                if self._pruning_mode == 'channel' and model_name in ['MobileNetV2', 'RetinaFace'] and name not in ops_no_dependency:
```
Comment (chicm-ms) [inline src/sdk/pynni/nni/compression/torch/pruning/net_adapt_pruner.py:105]: This checking seems very specific, would you explain the reason to do this check?

Comment (suiguoxin) [reply src/sdk/pynni/nni/compression/torch/pruning/net_adapt_pruner.py:105]: The PR https://github.com/microsoft/nni/pull/2435 provides functionalities to check dependencies. The checking here will be replaced after this PR is merged.

File: docs/en_US/Compressor/Pruner.md (hunk 1 above)
Comment (suiguoxin) [reply docs/en_US/Compressor/Pruner.md:553]: Description added.

File: src/sdk/pynni/nni/compression/torch/pruning/admm_pruner.py (hunk 3 above)
Comment (suiguoxin) [reply src/sdk/pynni/nni/compression/torch/pruning/admm_pruner.py]: Refactored with `Masker`.

File: src/sdk/pynni/nni/compression/torch/pruning/net_adapt_pruner.py (hunk 4 above)
Comment (suiguoxin) [reply src/sdk/pynni/nni/compression/torch/pruning/net_adapt_pruner.py:81]: I prefer to refactor this in another PR, together with some related issues, e.g. the calc of overall sparsity.

File: src/sdk/pynni/nni/compression/torch/pruning/net_adapt_pruner.py (hunk 5 above)
Comment (suiguoxin) [reply src/sdk/pynni/nni/compression/torch/pruning/net_adapt_pruner.py]: Yes, this is a solution recommended in the PyTorch forum to copy a model.

File: src/sdk/pynni/nni/compression/torch/pruning/net_adapt_pruner.py (hunk 6 above)
Comment (suiguoxin) [reply src/sdk/pynni/nni/compression/torch/pruning/net_adapt_pruner.py]: The checking is removed. To make it clear, I'll deal with this functionality (prune only ops without dependency) in a seperate PR in the future.

File: src/sdk/pynni/nni/compression/torch/pruning/admm_pruner.py (hunk 2 above)
Comment (suiguoxin) [reply src/sdk/pynni/nni/compression/torch/pruning/admm_pruner.py]: Fixed in validate_config.

**Comment by suiguoxin** [comment]
Changes since last review meeting:
- Resolve confict after # PR https://github.com/microsoft/nni/pull/2504
- Refactor ADMMPruner with `Masker` : not convenient to use `Masker` in other auto-pruners (ADMMPruner is not an auto-pruner, it does not deal with automatic sparsity distribution)
- Rename of some common class arguments:
    - Pruning_mode (channel, fine_grained)-> base_algo (level, l1, l2)
    - Fine_tuner -> short_term_fine_tune 
    -  Pruning_step -> sparsity_per_iteration (NetAdaptPruner)
    - other specific args

- Detailed explanation of function (args, return value):
    - evaluator
    - short_term_fine_tuner
    - trainer (ADMMPruner, AutoCompressPruner)
- Include l2 as an option of base_pruning_algo
- Remove check for specific models in pruner implementation
- NetAdaptPruner:
    - Function Check related ops: Fix known issues for some cases: 'dropout'/'pooling'… etc


Not done:
- Refactor auto-pruners(SAPruner/NetAdaptPruner) with `Masker`:
    - Not convenient to use `Masker` class to generate masks for all the ops at once given the config_list
-  Remove `speedup` from AutoCompressPruner:
    - This will involves huge changes to SimulatedAnnealing & ADMMPruner and make them difficult to maintain. As the `speedup` component is always necessary for application and we are supporting more and more models in speedup, I suggest to keep it for the moment.

File: src/sdk/pynni/nni/compression/torch/pruning/admm_pruner.py [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+        Parameters
+        ----------
+        wrapper : Module
+            the module to instrument the compression operation
+        Returns
+        -------
+        dict
+            dictionary for storing masks
+        """
+        self.masker = MASKER_DICT[self._base_algo](self.bound_model, self)
```
Comment (chicm-ms) [inline src/sdk/pynni/nni/compression/torch/pruning/admm_pruner.py:137]: This `self.masker` is not used.

Comment (suiguoxin) [reply src/sdk/pynni/nni/compression/torch/pruning/admm_pruner.py:137]: > This `self.masker` is not used.
`self.masker` is required by OneShotPruner to realize `calc_mask`

Comment (suiguoxin) [reply src/sdk/pynni/nni/compression/torch/pruning/admm_pruner.py]: Moved to `__init__` .

File: docs/en_US/Compressor/Pruner.md [hunk 8]
```diff
@@ -390,5 +395,170 @@ You can view example for more information
 - **op_types:** Currently only Conv2d is supported in TaylorFOWeightFilterPruner.
 
 
+## NetAdapt Pruner
+NetAdapt allows a user to automatically simplify a pretrained network to meet the resource budget. 
+For example, given the ovrall sparsity, NetAdapt will automatically generate the sparsities distribution among different layers by iterative pruning.
```
Comment (QuanluZhang) [inline docs/en_US/Compressor/Pruner.md:400]: ovrall -> overall

File: docs/en_US/Compressor/Pruner.md [hunk 9]
```diff
@@ -390,5 +395,170 @@ You can view example for more information
...
+pruner = NetAdaptPruner(model, config_list, short_term_fine_tuner=short_term_fine_tuner, evaluator=evaluator,base_algo='l1', experiment_data_dir='./')
+pruner.compress()
+```
+
+You can view [example](https://github.com/microsoft/nni/blob/master/examples/model_compress/auto_pruners_torch.py) for more information.
+
+#### User configuration for NetAdapt Pruner
+
+- **sparsity:** How much percentage of convolutional filters are to be pruned.
+- **op_types:** Currently only Conv2d is supported in TaylorFOWeightFilterPruner.
```
Comment (QuanluZhang) [inline docs/en_US/Compressor/Pruner.md:425]: in TaylorFOWei...?

File: docs/en_US/Compressor/Pruner.md [hunk 10]
```diff
@@ -390,5 +395,170 @@ You can view example for more information
...
+- **short_term_fine_tuner:** Function to short-term fine tune the masked model.
+This function should include `model` as the only parameter, and fine tune the model for a short term after each pruning iteration.
+- **evaluator:** Function to evaluate the masked model. This function should include `model` as the only parameter, and returns a scalar value.
+- **base_algo:** base pruning algorithm. 'level', 'l1' or 'l2', by default 'l1'.
+- **sparsity_per_iteration:** sparsity to prune in each iteration
```
Comment (QuanluZhang) [inline docs/en_US/Compressor/Pruner.md:430]: better to give more explanation on this argument. give the final sparsity 0.8. if sparsity_per_iteration is 01. Do you mean the total sparsity would be 0, 0.1, 0.2, 0.3, ..., 0.8?

File: docs/en_US/Compressor/Pruner.md [hunk 11]
```diff
@@ -390,5 +395,170 @@ You can view example for more information
...
+
+You can view [example](https://github.com/microsoft/nni/blob/master/examples/model_compress/auto_pruners_torch.py) for more information.
+
+#### User configuration for ADMM Pruner
+
+- **sparsity:** How much percentage of convolutional filters are to be pruned.
+- **op_types:** Currently only Conv2d is supported in ADMMPruner.
+- **trainer:** Function used for the first optimization subproblem.
+This function should include `model, optimizer, criterion, epoch, callback` as parameters, where callback should be inserted after loss.backward of the normal training process.
+- **optimize_iteration:** ADMM optimize iterations.
```
Comment (QuanluZhang) [inline docs/en_US/Compressor/Pruner.md:513]: you mean the number of iterations?

File: docs/en_US/Compressor/Pruner.md [hunk 12]
```diff
@@ -390,5 +395,170 @@ You can view example for more information
...
+- **training_epochs:** training epochs of the first optimization subproblem.
```
Comment (QuanluZhang) [inline docs/en_US/Compressor/Pruner.md:514]: it is not clear what is "the first optimization subproblem". better to give a little more description in the introduction of this pruner

File: docs/en_US/Compressor/Pruner.md [hunk 13]
```diff
@@ -390,5 +395,170 @@ You can view example for more information
...
+- **row:** penalty parameters for ADMM training.
+- **base_algo:** base pruning algorithm. 'level', 'l1' or 'l2', by default 'l1'.
```
Comment (QuanluZhang) [inline docs/en_US/Compressor/Pruner.md:516]: why this one does not have `experiment_data_dir`?

File: docs/en_US/Compressor/Pruner.md [hunk 14]
```diff
@@ -390,5 +395,170 @@ You can view example for more information
...
+            
+
+## AutoCompress Pruner
+For each round t, AutoCompressPruner prune the model for the same sparsity each round to achive the ovrall sparsity:
```
Comment (QuanluZhang) [inline docs/en_US/Compressor/Pruner.md:520]: ovrall -> overall

File: docs/en_US/Compressor/Pruner.md [hunk 15]
```diff
@@ -390,5 +395,170 @@ You can view example for more information
...
+
+You can view [example](https://github.com/microsoft/nni/blob/master/examples/model_compress/auto_pruners_torch.py) for more information.
 
- 
\ No newline at end of file
+#### User configuration for AutoCompress Pruner
+
+- **sparsity:** How much percentage of convolutional filters are to be pruned.
+- **op_types:** "Conv2d" or "default".
+- **trainer:** Function used for the first optimization subproblem.
```
Comment (QuanluZhang) [inline docs/en_US/Compressor/Pruner.md:551]: it is not clear how to write `trainer`. who should provide `callback`? what is the reason to provide `callback`? why it should be put behind `loss.backward`?

File: docs/en_US/Compressor/Pruner.md [hunk 16]
```diff
@@ -390,5 +395,170 @@ You can view example for more information
...
+This function should include `model, optimizer, criterion, epoch, callback` as parameters, where callback should be inserted after loss.backward of the normal training process.
+- **evaluator:** Function to evaluate the masked model. This function should include `model` as the only parameter, and returns a scalar value.
+- **dummy_input:** The dummy input for model speed up, users should put it on right device before pass in.
```
Comment (QuanluZhang) [inline docs/en_US/Compressor/Pruner.md:554]: why there is model speed up here?

File: docs/en_US/Compressor/Pruner.md [hunk 17]
```diff
@@ -390,5 +395,170 @@ You can view example for more information
...
+- **iterations:** The number of overall iterations.
+- **optimize_mode:** Optimize mode, 'maximize' or 'minimize', by default 'maximize'.
```
Comment (QuanluZhang) [inline docs/en_US/Compressor/Pruner.md:556]: only this auto pruner supports `optimize_mode`?

File: docs/en_US/Compressor/Pruner.md [hunk 18]
```diff
@@ -390,5 +395,170 @@ You can view example for more information
...
+- **base_algo:** base pruning algorithm. 'level', 'l1' or 'l2', by default 'l1'.
+- **start_temperature:** Simualated Annealing related parameter.
+- **stop_temperature:** Simualated Annealing related parameter.
+- **cool_down_rate:** Simualated Annealing related parameter.
+- **perturbation_magnitude:** Initial perturbation magnitude to the sparsities. The magnitude decreases with current temperature.
+- **optimize_iteration:** ADMM optimize iterations.
```
Comment (QuanluZhang) [inline docs/en_US/Compressor/Pruner.md:562]: what is the relation with ADMM?

File: docs/en_US/Compressor/Pruner.md [hunk 19]
```diff
@@ -390,5 +395,170 @@ You can view example for more information
...
+- **epochs:** training epochs of the first optimization subproblem.
```
Comment (QuanluZhang) [inline docs/en_US/Compressor/Pruner.md:563]: this one also has two subproblems?

File: docs/en_US/Compressor/Pruner.md (hunk 8 above)
Comment (suiguoxin) [reply docs/en_US/Compressor/Pruner.md]: Fixed

File: docs/en_US/Compressor/Pruner.md (hunk 9 above)
Comment (suiguoxin) [reply docs/en_US/Compressor/Pruner.md]: Fixed

File: docs/en_US/Compressor/Pruner.md (hunk 10 above)
Comment (suiguoxin) [reply docs/en_US/Compressor/Pruner.md]: Yes. More explanation added.

File: docs/en_US/Compressor/Pruner.md (hunk 11 above)
Comment (suiguoxin) [reply docs/en_US/Compressor/Pruner.md]: Yes and renamed.

File: docs/en_US/Compressor/Pruner.md (hunk 12 above)
Comment (suiguoxin) [reply docs/en_US/Compressor/Pruner.md]: Added.

File: docs/en_US/Compressor/Pruner.md (hunk 13 above)
Comment (suiguoxin) [reply docs/en_US/Compressor/Pruner.md]: ADMMPruner is not an auto pruner, there is thus no experiment data generated. More explanation on what is included as experiment data added for the auto pruners.

File: docs/en_US/Compressor/Pruner.md (hunk 14 above)
Comment (suiguoxin) [reply docs/en_US/Compressor/Pruner.md]: Fixed

File: docs/en_US/Compressor/Pruner.md (hunk 17 above)
Comment (suiguoxin) [reply docs/en_US/Compressor/Pruner.md]: `optimize_mode` is supported in `NetAdapt` `SimualatedAnnealing` and `AucoCompress`. Sorry for have missed this arg for `NetAdaptPruner`.

File: docs/en_US/Compressor/Pruner.md (hunk 18 above)
Comment (suiguoxin) [reply docs/en_US/Compressor/Pruner.md]: AutoCompress Pruner call `SimualtedAnnealing Pruner`  and `ADMM Pruner ` iteratively.

File: docs/en_US/Compressor/Pruner.md (hunk 19 above)
Comment (suiguoxin) [reply docs/en_US/Compressor/Pruner.md]: These are args for `ADMM`

File: docs/en_US/Compressor/Pruner.md (hunk 16 above)
Comment (suiguoxin) [reply docs/en_US/Compressor/Pruner.md:552]: `speedup` is called inside `AutoCompress` to keep the model un-masked and realize real pruning after each iteration.

File: docs/en_US/Compressor/Pruner.md (hunk 15 above)
Comment (suiguoxin) [reply docs/en_US/Compressor/Pruner.md]: updated

File: src/sdk/pynni/nni/compression/torch/pruning/net_adapt_pruner.py [hunk 20]
```diff
@@ -0,0 +1,324 @@
...
+        evaluator : function
+            function to evaluate the masked model.
+            This function should include `model` as the only parameter, and returns a scalar value.
+        short_term_fine_tuner : function
+            function to short-term fine tune the masked model.
+            This function should include `model` as the only parameter,
+            and fine tune the model for a short term after each pruning iteration.
+        optimize_mode : str
+            optimize mode, 'maximize' or 'minimize', by default 'maximize'
+        base_algo : str
```
[the remark this answers is not in the document]
Comment (suiguoxin) [reply src/sdk/pynni/nni/compression/torch/pruning/net_adapt_pruner.py:59]: added

File: src/sdk/pynni/nni/compression/torch/pruning/net_adapt_pruner.py [hunk 21]
```diff
@@ -0,0 +1,324 @@
...
+class NetAdaptPruner(Pruner):
+    """
+    This is a Pytorch implementation of NetAdapt compression algorithm.
+
+    The pruning procedure can be described as follows:
+    While Res_i > Bud:
+        1. Con = Res_i - delta_Res
+        2. for every layer:
+            Choose Num Filters to prune
+            Choose which filter to prunee
```
[the remark this answers is not in the document]
Comment (suiguoxin) [reply src/sdk/pynni/nni/compression/torch/pruning/net_adapt_pruner.py]: fixed

File: src/sdk/pynni/nni/compression/torch/pruning/auto_compress_pruner.py [hunk 22]
```diff
@@ -0,0 +1,221 @@
...
+        Compress the model with AutoCompress.
+
+        Returns
+        -------
+        torch.nn.Module
+            model with specified modules compressed.
+        """
+        _logger.info('Starting AutoCompress pruning...')
+
+        sparsity_each_round = 1 - pow(1-self._sparsity, 1/self._optimize_iterations)
```
[the remark this answers is not in the document]
Comment (suiguoxin) [reply src/sdk/pynni/nni/compression/torch/pruning/auto_compress_pruner.py]: This strategy is used to ensure that same number of weights will be pruned in each iteration.

File: docs/en_US/Compressor/Pruner.md [hunk 23]
```diff
@@ -398,5 +402,176 @@ We try to reproduce the experiment result of the fully connected network on MNIS
 The above figure shows the result of the fully connected network. `round0-sparsity-0.0` is the performance without pruning. Consistent with the paper, pruning around 80% also obtain similar performance compared to non-pruning, and converges a little faster. If pruning too much, e.g., larger than 94%, the accuracy becomes lower and convergence becomes a little slower. A little different from the paper, the trend of the data in the paper is relatively more clear.
 
 
+## NetAdapt Pruner
```
Comment (chicm-ms) [inline docs/en_US/Compressor/Pruner.md:405]: The order of each section is better consistent with the content directory at beginning.

Comment (suiguoxin) [reply docs/en_US/Compressor/Pruner.md:356]: fixed

File: src/sdk/pynni/nni/compression/torch/pruning/auto_compress_pruner.py [hunk 24]
```diff
@@ -0,0 +1,227 @@
...
+                row=self._row,
+                base_algo=self._base_algo)
+            ADMMpruner.compress()
+
+            ADMMpruner.export_model(os.path.join(self._experiment_data_dir, 'model_admm_masked.pth'), os.path.join(
+                self._experiment_data_dir, 'mask.pth'))
+
+            # use speed up to prune the model before next iteration, because SimulatedAnnealingPruner & ADMMPruner don't take masked models
+            self._model_to_prune.load_state_dict(torch.load(os.path.join(
+                self._experiment_data_dir, 'model_admm_masked.pth')))
```
Comment (colorjam) [inline src/sdk/pynni/nni/compression/torch/pruning/auto_compress_pruner.py:195]: Why reload the checkpoint?

File: src/sdk/pynni/nni/compression/torch/pruning/admm_pruner.py [hunk 25]
```diff
@@ -0,0 +1,181 @@
...
+            users are just required to insert `callback()` between `loss.backward()` and `optimizer.step()`.
+        num_iterations : int
+            Total number of iterations.
+        training_epochs : int
+            Training epochs of the first subproblem.
+        row : float
+            Penalty parameters for ADMM training.
+        base_algo : str
+            Base pruning algorithm. `level`, `l1` or `l2`, by default `l1`.
+            Given the sparsity distrution among the ops, the assigned `base_algo` is used to decide which filters/channels/weights to prune.
```
Comment (colorjam) [inline src/sdk/pynni/nni/compression/torch/pruning/admm_pruner.py:51]: distrution -> distribution

Comment (suiguoxin) [reply src/sdk/pynni/nni/compression/torch/pruning/admm_pruner.py]: fixed

File: src/sdk/pynni/nni/compression/torch/pruning/auto_compress_pruner.py (hunk 24 above)
Comment (suiguoxin) [reply src/sdk/pynni/nni/compression/torch/pruning/auto_compress_pruner.py:227]: The model weights have changed after admm pruner.

**Pull request merged**